imagepuller/internal/service/verify_test.go

```diff
@@ -102,3 +118,225 @@ func TestStoreAndVerifyLayers_EvilRegistry(t *testing.T) {
+	}
+	return s.DefaultRemote.Head(ref, opts...)
+}
+
+func (s StubRemote) Image(ref name.Reference, opts ...remote.Option) (gcr.Image, error) {
+	if s.imageErr {
+		return nil, errors.New("")
+	}
+	return s.DefaultRemote.Image(ref, opts...)
+}
+
+func (s StubRemote) Index(ref name.Reference, opts ...remote.Option) (gcr.ImageIndex, error) {
+	if s.indexErr {
+		return nil, errors.New("")
+	}
+	return s.DefaultRemote.Index(ref, opts...)
+}
+
+func TestGetAndVerifyImage(t *testing.T) {
+	tests := map[string]struct {
+		digest        string
+		imageURLNoTag bool
+		wantErr       error
+		startRegistry bool
+		stubRemote    StubRemote
+	}{
+		"digest missing, no tag": {
+			digest:        "",
+			imageURLNoTag: true,
+			wantErr:       ErrParseDigest,
+		},
+		"digest malformed, no tag": {
+			digest:        "sha256:000",
+			imageURLNoTag: true,
+			wantErr:       ErrParseDigest,
+		},
+		"digest missing algorithm, no tag": {
+			digest:        "0000000000000000000000000000000000000000000000000000000000000000",
+			imageURLNoTag: true,
+			wantErr:       ErrParseDigest,
+		},
+		"digest missing, with tag": {
+			digest:  "",
+			wantErr: ErrParseDigest,
+		},
+		"digest malformed, with tag": {
+			digest:  "sha256:000",
+			wantErr: ErrParseDigest,
+		},
+		"digest missing algorithm, with tag": {
+			digest:  "0000000000000000000000000000000000000000000000000000000000000000",
+			wantErr: ErrParseDigest,
+		},
+		"head missing for valid, but non-existent digest": {
+			digest:        "sha256:0000000000000000000000000000000000000000000000000000000000000000",
+			wantErr:       ErrDescriptor,
+			stubRemote:    StubRemote{headErr: true},
+			startRegistry: true,
+		},
+		"head missing": {
+			digest:        registry.ManifestDigest(),
+			wantErr:       ErrDescriptor,
+			stubRemote:    StubRemote{headErr: true},
+			startRegistry: true,
+		},
+		"remote image failure": {
+			digest:        registry.ManifestDigest(),
+			wantErr:       ErrRemoteImage,
+			stubRemote:    StubRemote{imageErr: true},
+			startRegistry: true,
+		},
+		"remote index failure": {
+			digest:        registry.IndexDigest(),
+			wantErr:       ErrRemoteIndex,
+			stubRemote:    StubRemote{indexErr: true},
+			startRegistry: true,
+		},
+		"unexpected media type": {
+			digest:        registry.UnknownMediaTypeDigest(),
+			wantErr:       ErrUnexpectedMediaType,
+			startRegistry: true,
+		},
+		"index valid, linux/amd64 missing": {
+			digest:        registry.IndexForMissingPlatformDigest(),
+			wantErr:       ErrMissingPlatform,
+			startRegistry: true,
+		},
+		"image success": {
+			digest:        registry.ManifestDigest(),
+			startRegistry: true,
+		},
+		"index success": {
+			digest:        registry.IndexDigest(),
+			startRegistry: true,
+		},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			assert := assert.New(t)
+			log := slog.Default()
+
+			var imageURL string
+			if tc.imageURLNoTag {
+				imageURL = fmt.Sprintf("@%s", tc.digest)
```

Comment:
In combination with `tc.startRegistry`, the resulting reference looks invalid: `localhost:123/@sha256:...`.